cli/login.go

```diff
@@ -1,91 +1,61 @@
-	scanner := bufio.NewScanner(r)
-	fmt.Printf("username: ")
-	username := ""
-	if scanner.Scan() {
-		username = scanner.Text()
-	} else {
-		return "", "", errUnableToReadUsername
-	}
-
-	fmt.Printf("password: ")
-	bytes, err := terminal.ReadPassword(int(syscall.Stdin))
-	if err != nil {
-		return "", "", fmt.Errorf("unable to get password: %w", err)
-	}
-
-	password := string(bytes)
-	// Need to add our own newline
-	fmt.Println()
-	return username, password, nil
-}
-
-func promptForCredentials(r io.Reader) (core.Credentials, error) {
-	username, password, err := getUsernameAndPassword(r)
-	return core.Credentials{Username: username, Password: password}, err
-}
-
-func init() {
-	loginCmd.Flags().StringVar(&identityProvider, "identity-provider", providers.Okta, "The identity provider to use.")
-}
+var (
+	ClientID   = os.Getenv("OKTA_CLIENT_ID")
+	OktaDomain = os.Getenv("OKTA_DOMAIN")
+)
 
 var loginCmd = &cobra.Command{
 	Use:   "login",
 	Short: "Authenticate with KeyConjurer.",
-	Long:  "Login using your AD creds. This stores encrypted credentials on the local system.",
+	Long:  "Login to KeyConjurer using OAuth2. You will be required to open the URL printed to the console or scan a QR code.",
 	// Example: appname + " login",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		ctx := context.Background()
-		client, err := newClient()
-		if err != nil {
-			return err
+		if !HasTokenExpired(config.Tokens) {
+			return nil
 		}
 
-		creds, err := promptForCredentials(os.Stdin)
+		token, err := Login(cmd.Context(), OktaDomain, false)
 		if err != nil {
 			return err
 		}
 
-		data, err := client.GetUserData(ctx, &GetUserDataOptions{
-			Credentials:            creds,
-			AuthenticationProvider: identityProvider,
-		})
+		return config.SaveOAuthToken(token)
+	},
+}
 
-		if err != nil {
-			return err
-		}
+func Login(ctx context.Context, domain string, useDeviceFlow bool) (*oauth2.Token, error) {
+	oauthCfg, provider, err := DiscoverOAuth2Config(ctx, domain)
+	if err != nil {
+		return nil, err
+	}
 
-		config.Creds = data.EncryptedCredentials
-		var entries []Account
-		for _, acc := range data.Apps {
-			entries = append(entries, Account{ID: acc.ID, Name: acc.Name, Alias: generateDefaultAlias(acc.Name)})
-		}
+	state, err := GenerateState()
+	if err != nil {
+		return nil, err
+	}
 
-		config.Accounts.ReplaceWith(entries)
-		if !quiet {
-			fmt.Fprintf(os.Stderr, "logged into %q successfully\n", creds.Username)
-		}
-		return nil
-	},
+	codeVerifier, codeChallenge, err := GenerateCodeVerifierAndChallenge()
+	if err != nil {
+		return nil, err
+	}
+
+	// The device flow and the redirect flow are almost indistinguishable from a user point of view.
+	//
+	// The device flow should be preferred as it gives the user the option to open a browser on their mobile device or their terminal, whereas the redirect flow requires opening a browser on the current machine.
+	if useDeviceFlow && oidc.SupportsDeviceFlow(provider) {
```

Comment:
It would be really cool to re-implement this but as things stand it doesn't look like we can get an id_token from the device auth flow